{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SATIVA requires a 'taxonomy file', i.e. a text file that links each sequence record to a ncbi taxonomic id (aka taxid - see here).\n",
    "\n",
    "We are going to generate this file from a bunch of sequence records downloaded from Genbank.\n",
    "\n",
    "We have downloaded our reference sequences from Genbank in the so-called, genbank format, which is a rich format containing all sorts of metadata for every sequence record, including a specific taxonomic id (taxid), i.e. a identity number that uniquely identifies every species (and also higher taxonomic levels, such as genera, families, classes, etc) recorded on Genbank.\n",
    "\n",
    "Let's parse our Genbank file using Biopython.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "records = SeqIO.to_dict(SeqIO.parse(open('../1-fetch_clean_align/12S_preSATIVA_fish.gb','r'),'genbank'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the below cell we display the first of our sequence records in Genbank format. Try to identify the taxid of the record and use NCBI's taxonomy database (here) to retrieve the full taxonomic tree for the taxon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       KC292935                1462 bp    DNA     linear   VRT 13-JUN-2013\n",
      "DEFINITION  Cyprinus carpio isolate WKL42 tRNA-Pro gene, D-loop, and tRNA-Phe\n",
      "            gene, complete sequence; and 12S ribosomal RNA gene, partial\n",
      "            sequence; mitochondrial.\n",
      "ACCESSION   KC292935\n",
      "VERSION     KC292935.1\n",
      "KEYWORDS    .\n",
      "SOURCE      mitochondrion Cyprinus carpio (common carp)\n",
      "  ORGANISM  Cyprinus carpio\n",
      "            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\n",
      "            Actinopterygii; Neopterygii; Teleostei; Ostariophysi; Cypriniformes;\n",
      "            Cyprinidae; Cyprinus.\n",
      "REFERENCE   1  (bases 1 to 1462)\n",
      "  AUTHORS   Hao,J., Yang,Q., Bao,D., Liang,A.-J., Zhang,X.-H. and Dong,S.\n",
      "  TITLE     The sequence comparison of mtDNA D-loop and adjacent regions in six\n",
      "            fish species\n",
      "  JOURNAL   Dalian Hai Yang Da Xue Xue Bao 28 (2), 160-165 (2013)\n",
      "REFERENCE   2  (bases 1 to 1462)\n",
      "  AUTHORS   Hao,J., Dong,S., Yang,Q., Bao,D., Liang,A. and Zhang,X.\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (03-DEC-2012) College of Life Science, Tianjin Normal\n",
      "            University, No 393, Binshuixi Road, Xiqing, Tianjin, Tianjin 300387,\n",
      "            China\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..1462\n",
      "                     /db_xref=\"taxon:7962\"\n",
      "                     /feature_id=\"KC292935.1_source\"\n",
      "                     /isolate=\"WKL42\"\n",
      "                     /mol_type=\"genomic DNA\"\n",
      "                     /organelle=\"mitochondrion\"\n",
      "                     /organism=\"Cyprinus carpio\"\n",
      "     rRNA            1068..>1462\n",
      "                     /GC_content=\"49.3670886076\"\n",
      "                     /feature_id=\"KC292935.1_f0\"\n",
      "                     /nuc_degen_prop=\"0.0\"\n",
      "                     /product=\"12S ribosomal RNA\"\n",
      "ORIGIN\n",
      "        1 ccagaaaaga gagattttaa ctctcacccc tggctcccaa agccagaatt ctaaactaaa\n",
      "       61 ctattttctg gggaggtagc actcccttta tggtatagta catattatgc ataatattac\n",
      "      121 attaatgtat tagtacatat atgtattatc accaactcac tattttaacc ataaagcagg\n",
      "      181 tacataatat taaggtgggc ataaagcata ttattaagac tcacaaattc tattatttgg\n",
      "      241 acttgagtaa tatattaatc cccaaaaatt tgtcctcaaa tttttccttg aaataatcaa\n",
      "      301 ctataattcc atttaaacat attaatgtag taagagacca ccaaccagtt tatataaagg\n",
      "      361 tatatcatga atgatagaat caaggacaat aattgtgagg gttacacaat atgaactatt\n",
      "      421 actggcatct ggttcctatt tcagggacat aactgtaata ttccaccctc ggataattat\n",
      "      481 actggcatct gattaatggt gtaagtacat acgtttcatt accccccatg ccgagcattc\n",
      "      541 ttttatatgc ataaggtatt ttttattggt ttcctttcat ctggcatttc agagtgcagg\n",
      "      601 ctcaaatgtt aaatcaaggt tgaacatttt ccttgtatgt gataatatat attaattatc\n",
      "      661 gtaagacata atttaagaat tacatacttt tatctcaagt gcataatata tctgtctcta\n",
      "      721 gttcaactta tccttacata gtgccccctt tggtttttgc gcgacaaacc cccttacccc\n",
      "      781 ctacgctcag cgaatcctgt tatccttgtc aaaccccgaa accaaggagg acccaagaac\n",
      "      841 gtgtaaacca acgagttgag gtataaattg gcatcccatt atatatatat atatatatgc\n",
      "      901 atcggttttt ttaaccgcaa cttaccactt acctaaaagc ccctaccaaa aatccccaaa\n",
      "      961 aagaggctcg gcactaaata ctctaatata attaatcagc tagcgtagct taacacaaag\n",
      "     1021 catagcactg aagatgctaa gatgagtcct aaaaaactcc gcatgcacaa aggcatggtc\n",
      "     1081 ccgaccttat tatcagctct aactcaactt acacatgcaa gtctccgcaa cccagtgaat\n",
      "     1141 atgccctcaa tccccccacc cggggacgag gagcgggcat caggcacaaa cattagccca\n",
      "     1201 agacgcctag ccaagccaca cccccaaggg aattcagcag tgataaacat taagccataa\n",
      "     1261 gtgaaaactt gactcagtta gtgttaagag ggccggtaaa actcgtgcca gccaccgcgg\n",
      "     1321 ttagacgaga ggccctagtt gatattacaa cggcgtaaag ggtggttaag gataaacaaa\n",
      "     1381 aataaagtca aatggcccct tggccgtcat acgcttctag gagtccgaag ccctaatacg\n",
      "     1441 aaagtaactt aataaaccca cc\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for r in records.keys():\n",
    "    print records[r].format('genbank')\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We don't want to make our life harder than it already is and deal with subspecies so in the tax file for SATIVA we'll only do species.\n",
    "\n",
    "Iterate over all records and check if taxid is subspecies, if yes replace with taxid of species and extract taxonomy line. If it's a subspecies record the record id for further processing.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cyprinus carpio  .. add to records - 7962\n",
      "Perca fluviatilis  .. add to records - 8168\n",
      "Cobitis taenia  .. add to records - 98395\n",
      "subspecies: Cyprinus carpio haematopterus\n",
      "Pungitius pungitius  .. add to records - 134920\n",
      "Solea solea  .. add to records - 90069\n",
      "subspecies: Gasterosteus aculeatus williamsoni\n",
      "Esox lucius  .. add to records - 8010\n",
      "Cobitis taenia  .. already covered\n",
      "Hypophthalmichthys molitrix  .. add to records - 13095\n",
      "Oncorhynchus mykiss  .. add to records - 8022\n",
      "Rhodeus sericeus  .. add to records - 58327\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Phoxinus phoxinus  .. add to records - 58324\n",
      "Carassius auratus  .. add to records - 7957\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Salmo salar  .. add to records - 8030\n",
      "Ambloplites rupestris  .. add to records - 109273\n",
      "Coregonus lavaretus  .. add to records - 59291\n",
      "Alburnus alburnus  .. add to records - 54556\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Acipenser sturio  .. add to records - 61674\n",
      "Anguilla anguilla  .. add to records - 7936\n",
      "Hypophthalmichthys nobilis  .. add to records - 7965\n",
      "Alburnoides bipunctatus  .. add to records - 58321\n",
      "Gasterosteus aculeatus  .. add to records - 69293\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gymnocephalus cernua  .. add to records - 57865\n",
      "Pungitius pungitius  .. already covered\n",
      "Squalius cephalus  .. add to records - 8284\n",
      "Lepomis gibbosus  .. add to records - 270329\n",
      "Barbatula barbatula  .. add to records - 135647\n",
      "Coregonus albula  .. add to records - 36180\n",
      "Ameiurus nebulosus  .. add to records - 27778\n",
      "Pimephales promelas  .. add to records - 90988\n",
      "Coregonus autumnalis  .. add to records - 27773\n",
      "Platichthys flesus  .. add to records - 8260\n",
      "Squalius cephalus  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Tinca tinca  .. add to records - 27717\n",
      "Salmo trutta  .. add to records - 8032\n",
      "Pungitius pungitius  .. already covered\n",
      "Pseudorasbora parva  .. add to records - 51549\n",
      "Pseudorasbora parva  .. already covered\n",
      "Gobio gobio  .. add to records - 27704\n",
      "Gobio gobio  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Neogobius melanostomus  .. add to records - 47308\n",
      "Ameiurus nebulosus  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Ameiurus nebulosus  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Solea solea  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Blicca bjoerkna  .. add to records - 58317\n",
      "Blicca bjoerkna  .. already covered\n",
      "Ctenopharyngodon idella  .. add to records - 7959\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Micropterus salmoides  .. add to records - 27706\n",
      "subspecies: Carassius auratus auratus\n",
      "Ambloplites rupestris  .. already covered\n",
      "Leuciscus leuciscus  .. add to records - 58325\n",
      "Lota lota  .. add to records - 69944\n",
      "Misgurnus bipartitus  .. add to records - 341117\n",
      "Misgurnus bipartitus  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Leuciscus idus  .. add to records - 69811\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "subspecies: Carassius auratus ssp. 'Pingxiang'\n",
      "Misgurnus fossilis  .. add to records - 7984\n",
      "Misgurnus fossilis  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Salvelinus alpinus  .. add to records - 8036\n",
      "Petromyzon marinus  .. add to records - 7757\n",
      "Salmo salar  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Salvelinus fontinalis  .. add to records - 8038\n",
      "Cyprinus carpio  .. already covered\n",
      "Pomatoschistus minutus  .. add to records - 13225\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "subspecies: Gasterosteus aculeatus williamsoni\n",
      "Salmo salar  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Umbra pygmaea  .. add to records - 75934\n",
      "Carassius auratus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Cottus gobio  .. add to records - 100952\n",
      "Salvelinus alpinus  .. already covered\n",
      "Coregonus oxyrinchus  .. add to records - 348708\n",
      "Chondrostoma nasus  .. add to records - 52619\n",
      "Salvelinus alpinus  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Umbra pygmaea  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Abramis brama  .. add to records - 38527\n",
      "Abramis brama  .. already covered\n",
      "Solea solea  .. already covered\n",
      "Scardinius erythrophthalmus  .. add to records - 58319\n",
      "Barbatula barbatula  .. already covered\n",
      "Barbus barbus  .. add to records - 40830\n",
      "Barbatula barbatula  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "Squalius cephalus  .. already covered\n",
      "Leuciscus leuciscus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Ameiurus melas  .. add to records - 219545\n",
      "Tinca tinca  .. already covered\n",
      "Umbra pygmaea  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Solea solea  .. already covered\n",
      "Solea solea  .. already covered\n",
      "Silurus glanis  .. add to records - 94993\n",
      "Tinca tinca  .. already covered\n",
      "Thymallus thymallus  .. add to records - 36185\n",
      "Squalius cephalus  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Sander lucioperca  .. add to records - 283035\n",
      "subspecies: Salmo trutta macrostigma\n",
      "Coregonus lavaretus  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Osmerus eperlanus  .. add to records - 29151\n",
      "Leucaspius delineatus  .. add to records - 58323\n",
      "Alosa alosa  .. add to records - 278164\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Gymnocephalus cernua  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Pomatoschistus minutus  .. already covered\n",
      "Phoxinus phoxinus  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Rutilus rutilus  .. add to records - 48668\n",
      "Salmo trutta  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Gymnocephalus cernua  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "Lepomis gibbosus  .. already covered\n",
      "subspecies: Salmo trutta fario\n",
      "Cottus gobio  .. already covered\n",
      "Oncorhynchus gorbuscha  .. add to records - 8017\n",
      "Pimephales promelas  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "subspecies: Micropterus salmoides salmoides\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Rhodeus amarus  .. add to records - 98397\n",
      "Alburnus alburnus  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "subspecies: Salmo trutta fario\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Phoxinus phoxinus  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Thymallus thymallus  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Lota lota  .. already covered\n",
      "subspecies: Cyprinus carpio 'wuyuanensis'\n",
      "Esox lucius  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Gobio gobio  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Thymallus thymallus  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "subspecies: Carassius auratus ssp. 'Kinbuna'\n",
      "Pungitius pungitius  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Acipenser sturio  .. already covered\n",
      "subspecies: Abramis brama orientalis\n",
      "Pungitius pungitius  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Abramis brama  .. already covered\n",
      "Tinca tinca  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Rhodeus amarus  .. already covered\n",
      "Thymallus thymallus  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "Alosa fallax  .. add to records - 278165\n",
      "Scardinius erythrophthalmus  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "subspecies: Phoxinus phoxinus tumensis\n",
      "Anguilla anguilla  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Silurus glanis  .. already covered\n",
      "Salmo trutta  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Rhodeus amarus  .. already covered\n",
      "Rhodeus amarus  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Rhodeus sericeus  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Thymallus thymallus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "subspecies: Salmo trutta trutta\n",
      "Alburnus alburnus  .. already covered\n",
      "Alburnus alburnus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Barbatula barbatula  .. already covered\n",
      "subspecies: Cyprinus carpio 'color'\n",
      "Pomatoschistus minutus  .. already covered\n",
      "Pungitius pungitius  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Gasterosteus aculeatus  .. already covered\n",
      "Cottus gobio  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Rhodeus amarus  .. already covered\n",
      "Carassius carassius  .. add to records - 217509\n",
      "Platichthys flesus  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Cottus gobio  .. already covered\n",
      "Ameiurus melas  .. already covered\n",
      "Cottus gobio  .. already covered\n",
      "Cottus gobio  .. already covered\n",
      "Lota lota  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "subspecies: Carassius auratus grandoculis\n",
      "Micropterus salmoides  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Leuciscus idus  .. already covered\n",
      "Lota lota  .. already covered\n",
      "subspecies: Carassius auratus auratus\n",
      "Cyprinus carpio  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Pseudorasbora parva  .. already covered\n",
      "Ameiurus melas  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Barbus barbus  .. already covered\n",
      "Micropterus salmoides  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "subspecies: Cyprinus carpio 'wananensis'\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Gymnocephalus cernua  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Cobitis taenia  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Rutilus rutilus  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Salvelinus alpinus  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Lampetra fluviatilis  .. add to records - 7748\n",
      "Ponticola kessleri  .. add to records - 320565\n",
      "Pseudorasbora parva  .. already covered\n",
      "Carassius auratus  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Anguilla anguilla  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Oncorhynchus mykiss  .. already covered\n",
      "Salmo salar  .. already covered\n",
      "Coregonus lavaretus  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Hypophthalmichthys nobilis  .. already covered\n",
      "Osmerus eperlanus  .. already covered\n",
      "Neogobius fluviatilis  .. add to records - 320562\n",
      "Osmerus eperlanus  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Platichthys flesus  .. already covered\n",
      "Pomatoschistus minutus  .. already covered\n",
      "Pomatoschistus minutus  .. already covered\n",
      "Ponticola kessleri  .. already covered\n",
      "subspecies: Salmo trutta fario\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Cyprinus carpio  .. already covered\n",
      "Pimephales promelas  .. already covered\n",
      "Sander lucioperca  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n",
      "Esox lucius  .. already covered\n",
      "Perca fluviatilis  .. already covered\n",
      "Ctenopharyngodon idella  .. already covered\n",
      "Phoxinus phoxinus  .. already covered\n",
      "Hypophthalmichthys molitrix  .. already covered\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "taxon_to_taxid = {}\n",
    "recs_to_adjust = []\n",
    "taxon_to_recs = defaultdict(list)\n",
    "\n",
    "for key in records.keys():\n",
    "    r = records[key]\n",
    "   \n",
    "    source = [f for f in r.features if f.type == 'source'][0]\n",
    "    if (len(source.qualifiers['organism'][0].split(\" \")) == 2):\n",
    "        print source.qualifiers['organism'][0],\n",
    "        if 'db_xref' in source.qualifiers:\n",
    "#            print source.qualifiers['db_xref']\n",
    "            for t in source.qualifiers['db_xref']:\n",
    "#                print t\n",
    "                if 'taxon' in t:\n",
    "                    if not source.qualifiers['organism'][0] in taxon_to_taxid:\n",
    "                        print \" .. add to records - %s\" %t.split(\":\")[1]\n",
    "                        taxon_to_taxid[source.qualifiers['organism'][0]] = t.split(\":\")[1]\n",
    "                    else:\n",
    "                        print \" .. already covered\"\n",
    "                    taxon_to_recs[source.qualifiers['organism'][0]].append(r.id)\n",
    "    elif (len(source.qualifiers['organism'][0].split(\" \")) > 2) and ((source.qualifiers['organism'][0].split(\" \")[1] == 'sp.') or ('cf.' in source.qualifiers['organism'][0].split(\" \"))):\n",
    "        print source.qualifiers['organism'][0],\n",
    "        if 'db_xref' in source.qualifiers:\n",
    "            for t in source.qualifiers['db_xref']:\n",
    "                if 'taxon' in t:\n",
    "                    if not str(t) in taxon_to_taxid:\n",
    "                        taxon_to_taxid[source.qualifiers['organism'][0]] = t.split(\":\")[1]\n",
    "                    else:\n",
    "                        print \" .. already covered\" \n",
    "                    taxon_to_recs[source.qualifiers['organism'][0]].append(r.id)\n",
    "    else:\n",
    "        print \"subspecies: %s\" %source.qualifiers['organism'][0]\n",
    "        recs_to_adjust.append(r.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['JX188254.1', 'KM273828.1', 'AP011236.1', 'HQ875340.1', 'KM282418.1', 'EU048341.1', 'KP171723.1', 'HQ391896.1', 'GU233801.1', 'AB111951.1', 'JN105357.1', 'AB379915.1', 'LC069402.1', 'KC894466.1', 'KC992395.1', 'AM910409.1', 'JX188253.1', 'GU086397.1', 'AP011239.1', 'GU086395.1', 'KF856964.1', 'KT634053.1']\n"
     ]
    }
   ],
   "source": [
    "print recs_to_adjust"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the records that were identified as being *subspecies* reduce to species and check whether the taxid for the species had already been encountered before. If not we'd need to fetch it from NCBI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cyprinus carpio haematopterus -> Cyprinus carpio\n",
      "Gasterosteus aculeatus williamsoni -> Gasterosteus aculeatus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus ssp. 'Pingxiang' -> Carassius auratus\n",
      "Gasterosteus aculeatus williamsoni -> Gasterosteus aculeatus\n",
      "Salmo trutta macrostigma -> Salmo trutta\n",
      "Salmo trutta fario -> Salmo trutta\n",
      "Micropterus salmoides salmoides -> Micropterus salmoides\n",
      "Salmo trutta fario -> Salmo trutta\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Cyprinus carpio 'wuyuanensis' -> Cyprinus carpio\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus ssp. 'Kinbuna' -> Carassius auratus\n",
      "Abramis brama orientalis -> Abramis brama\n",
      "Phoxinus phoxinus tumensis -> Phoxinus phoxinus\n",
      "Salmo trutta trutta -> Salmo trutta\n",
      "Cyprinus carpio 'color' -> Cyprinus carpio\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Carassius auratus grandoculis -> Carassius auratus\n",
      "Carassius auratus auratus -> Carassius auratus\n",
      "Cyprinus carpio 'wananensis' -> Cyprinus carpio\n",
      "Salmo trutta fario -> Salmo trutta\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "to_fetch = defaultdict(list)\n",
    "\n",
    "for key in records.keys():\n",
    "    r = records[key]\n",
    "    if r.id in recs_to_adjust:\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        adjust_from = source.qualifiers['organism'][0]\n",
    "        adjust_to = \" \".join(adjust_from.split(\" \")[:2])\n",
    "        print \"%s -> %s\" %(adjust_from,adjust_to)\n",
    "        if adjust_to in taxon_to_taxid:\n",
    "            taxon_to_recs[adjust_to].append(r.id)\n",
    "        else:\n",
    "            to_fetch[adjust_to].append(r.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if we are good, or if there are any taxids missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Have taxids for all records\n"
     ]
    }
   ],
   "source": [
    "if to_fetch:\n",
    "    print \"need to fetch some taxids\"\n",
    "    print to_fetch\n",
    "else:\n",
    "    print \"Have taxids for all records\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If there are any taxids missing the next cell will fetch them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#from Bio import Entrez\n",
    "#Entrez.email = \"L.Harper@2015.hull.ac.uk\"\n",
    "\n",
    "#for binomial in to_fetch:\n",
    "#    print binomial\n",
    "#    handle = Entrez.esearch(db=\"Taxonomy\", term=binomial)\n",
    "#    record = Entrez.read(handle)\n",
    "#    print record[\"IdList\"][0]\n",
    "#    taxon_to_taxid[binomial] = record[\"IdList\"][0]\n",
    "    \n",
    "#    taxon_to_recs[binomial] = to_fetch[binomial]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write taxids to file and fetch full taxonomy for all of them using taxit from the taxtastic package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "taxids = []\n",
    "\n",
    "out=open(\"taxids.txt\",'w')\n",
    "for sp in taxon_to_taxid:\n",
    "    taxids.append(taxon_to_taxid[sp])\n",
    "    out.write(taxon_to_taxid[sp]+\"\\n\")\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An IndexError may be returned if record names do not match the taxonomy e.g. Lixus cf. pulverulentus AS-2013 instead of Lixus pulverulentus.\n",
    "\n",
    "If this occurs, you must go back to code cell 3 and change the code to split the record name where random letters occur (cf.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create comma-delimited text file with full taxonomic tree for each taxid: `taxa.csv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!taxit taxtable  /usr/bin/taxonomy.db -f taxids.txt -o taxa.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The metaBEAT image comes prepared with a database that contains taxonomic information for taxids. However, new taxids are constantly being added to Genbank, and the database (it's stored in the image in `/usr/bin/taxonomy.db`) contained in the image is not necessarily up to date. If the above command gives you errors about certain taxids not being found in the database you can try to update the database using code below. This will take a few minutes, depending on your internet connection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "downloading ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdmp.zip to /usr/bin/taxdmp.zip\n",
      "creating new database in /usr/bin/taxonomy.db using data in /usr/bin/taxdmp.zip\n",
      "0 records lack primary names\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Remove the old database\n",
    "rm /usr/bin/taxonomy.db\n",
    "\n",
    "# Create a new database. This command will first download the latest NCBI taxonomy info (taxdmp.zip) and configure the database\n",
    "taxit new_database -d /usr/bin/taxonomy.db\n",
    "\n",
    "# Remove temporal file that has been downloaded by the previous command\n",
    "rm /usr/bin/taxdmp.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, rerun code to create 'taxa.csv' file with full taxonomic tree for each taxid, using the latest taxonomy database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!taxit taxtable /usr/bin/taxonomy.db -f taxids.txt -o taxa.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The full 'taxonomy string' for a given taxon as returned from NCBI could look, e.g. like so:\n",
    "\n",
    "cellular organisms; Eukaryota; Opisthokonta; Metazoa; Eumetazoa; Bilateria; Deuterostomia; Chordata; Craniata; Vertebrata; Gnathostomata; Teleostomi; Euteleostomi; Actinopterygii; Actinopteri; Neopterygii; Teleostei; Osteoglossocephalai; Clupeocephala; Euteleosteomorpha; Protacanthopterygii; Salmoniformes; Salmonidae; Salmoninae; Salmo; Salmo trutta\n",
    "\n",
    "The number of taxonomic levels in the taxonomy string may vary between taxa. Some taxonomic groups are classified into relatively uncommon intermediate taxonomic levels, that may not exist for other taxa.\n",
    "\n",
    "In order to make our lives easier downstream we will limit ourselves only to a defined set of the most common taxonomic levels, that should be known for pretty much all taxa.\n",
    "\n",
    "  -  superkingdom\n",
    "  -  phylum\n",
    "  -  class\n",
    "  -  order\n",
    "  -  family\n",
    "  -  genus\n",
    "  -  species\n",
    "\n",
    "Extract 'taxonomy string' for a specific set of taxonomic levlels.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48668 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Rutilus;Rutilus rutilus\n",
      "8030 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Salmo;Salmo salar\n",
      "27704 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Gobio;Gobio gobio\n",
      "58325 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Leuciscus;Leuciscus leuciscus\n",
      "69944 Eukaryota;Chordata;Actinopteri;Gadiformes;Lotidae;Lota;Lota lota\n",
      "278164 Eukaryota;Chordata;Actinopteri;Clupeiformes;Clupeidae;Alosa;Alosa alosa\n",
      "54556 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Alburnus;Alburnus alburnus\n",
      "36185 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Thymallus;Thymallus thymallus\n",
      "7748 Eukaryota;Chordata;unknown;Petromyzontiformes;Petromyzontidae;Lampetra;Lampetra fluviatilis\n",
      "40830 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Barbus;Barbus barbus\n",
      "8010 Eukaryota;Chordata;Actinopteri;Esociformes;Esocidae;Esox;Esox lucius\n",
      "58317 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Blicca;Blicca bjoerkna\n",
      "38527 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Abramis;Abramis brama\n",
      "8017 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Oncorhynchus;Oncorhynchus gorbuscha\n",
      "135647 Eukaryota;Chordata;Actinopteri;Cypriniformes;Nemacheilidae;Barbatula;Barbatula barbatula\n",
      "7959 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Ctenopharyngodon;Ctenopharyngodon idella\n",
      "13225 Eukaryota;Chordata;Actinopteri;Gobiiformes;Gobiidae;Pomatoschistus;Pomatoschistus minutus\n",
      "27773 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Coregonus;Coregonus autumnalis\n",
      "7757 Eukaryota;Chordata;unknown;Petromyzontiformes;Petromyzontidae;Petromyzon;Petromyzon marinus\n",
      "320562 Eukaryota;Chordata;Actinopteri;Gobiiformes;Gobiidae;Neogobius;Neogobius fluviatilis\n",
      "320565 Eukaryota;Chordata;Actinopteri;Gobiiformes;Gobiidae;Ponticola;Ponticola kessleri\n",
      "90988 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Pimephales;Pimephales promelas\n",
      "58323 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Leucaspius;Leucaspius delineatus\n",
      "59291 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Coregonus;Coregonus lavaretus\n",
      "7984 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cobitidae;Misgurnus;Misgurnus fossilis\n",
      "8022 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Oncorhynchus;Oncorhynchus mykiss\n",
      "7962 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Cyprinus;Cyprinus carpio\n",
      "219545 Eukaryota;Chordata;Actinopteri;Siluriformes;Ictaluridae;Ameiurus;Ameiurus melas\n",
      "8038 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Salvelinus;Salvelinus fontinalis\n",
      "27717 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Tinca;Tinca tinca\n",
      "52619 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Chondrostoma;Chondrostoma nasus\n",
      "8168 Eukaryota;Chordata;Actinopteri;Perciformes;Percidae;Perca;Perca fluviatilis\n",
      "27778 Eukaryota;Chordata;Actinopteri;Siluriformes;Ictaluridae;Ameiurus;Ameiurus nebulosus\n",
      "69293 Eukaryota;Chordata;Actinopteri;Perciformes;Gasterosteidae;Gasterosteus;Gasterosteus aculeatus\n",
      "58324 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Phoxinus;Phoxinus phoxinus\n",
      "7965 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Hypophthalmichthys;Hypophthalmichthys nobilis\n",
      "8260 Eukaryota;Chordata;Actinopteri;Pleuronectiformes;Pleuronectidae;Platichthys;Platichthys flesus\n",
      "27706 Eukaryota;Chordata;Actinopteri;Centrarchiformes;Centrarchidae;Micropterus;Micropterus salmoides\n",
      "100952 Eukaryota;Chordata;Actinopteri;Perciformes;Cottidae;Cottus;Cottus gobio\n",
      "7957 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Carassius;Carassius auratus\n",
      "90069 Eukaryota;Chordata;Actinopteri;Pleuronectiformes;Soleidae;Solea;Solea solea\n",
      "61674 Eukaryota;Chordata;Actinopteri;Acipenseriformes;Acipenseridae;Acipenser;Acipenser sturio\n",
      "58321 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Alburnoides;Alburnoides bipunctatus\n",
      "58327 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Rhodeus;Rhodeus sericeus\n",
      "283035 Eukaryota;Chordata;Actinopteri;Perciformes;Percidae;Sander;Sander lucioperca\n",
      "13095 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Hypophthalmichthys;Hypophthalmichthys molitrix\n",
      "8284 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Squalius;Squalius cephalus\n",
      "75934 Eukaryota;Chordata;Actinopteri;Esociformes;Umbridae;Umbra;Umbra pygmaea\n",
      "98395 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cobitidae;Cobitis;Cobitis taenia\n",
      "7936 Eukaryota;Chordata;Actinopteri;Anguilliformes;Anguillidae;Anguilla;Anguilla anguilla\n",
      "36180 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Coregonus;Coregonus albula\n",
      "8036 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Salvelinus;Salvelinus alpinus\n",
      "217509 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Carassius;Carassius carassius\n",
      "109273 Eukaryota;Chordata;Actinopteri;Centrarchiformes;Centrarchidae;Ambloplites;Ambloplites rupestris\n",
      "98397 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Rhodeus;Rhodeus amarus\n",
      "58319 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Scardinius;Scardinius erythrophthalmus\n",
      "278165 Eukaryota;Chordata;Actinopteri;Clupeiformes;Clupeidae;Alosa;Alosa fallax\n",
      "57865 Eukaryota;Chordata;Actinopteri;Perciformes;Percidae;Gymnocephalus;Gymnocephalus cernua\n",
      "341117 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cobitidae;Misgurnus;Misgurnus bipartitus\n",
      "69811 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Leuciscus;Leuciscus idus\n",
      "348708 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Coregonus;Coregonus oxyrinchus\n",
      "47308 Eukaryota;Chordata;Actinopteri;Gobiiformes;Gobiidae;Neogobius;Neogobius melanostomus\n",
      "51549 Eukaryota;Chordata;Actinopteri;Cypriniformes;Cyprinidae;Pseudorasbora;Pseudorasbora parva\n",
      "134920 Eukaryota;Chordata;Actinopteri;Perciformes;Gasterosteidae;Pungitius;Pungitius pungitius\n",
      "29151 Eukaryota;Chordata;Actinopteri;Osmeriformes;Osmeridae;Osmerus;Osmerus eperlanus\n",
      "94993 Eukaryota;Chordata;Actinopteri;Siluriformes;Siluridae;Silurus;Silurus glanis\n",
      "270329 Eukaryota;Chordata;Actinopteri;Centrarchiformes;Centrarchidae;Lepomis;Lepomis gibbosus\n",
      "8032 Eukaryota;Chordata;Actinopteri;Salmoniformes;Salmonidae;Salmo;Salmo trutta\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "tax_levels=['superkingdom','phylum','class','order','family','genus','species']\n",
    "indices = []\n",
    "taxdict = defaultdict(list)\n",
    "taxids_to_taxonomy = {}\n",
    "\n",
    "infile=open(\"taxa.csv\",'r')\n",
    "header=infile.next()\n",
    "\n",
    "header_as_list=header.strip().replace('\"','').split(\",\")\n",
    "for i in range(len(header_as_list)):\n",
    "#    print header_as_list[i]\n",
    "    if header_as_list[i] in tax_levels:\n",
    "#        print \"\\t\"+header_as_list[i],i\n",
    "        indices.append(i)\n",
    "\n",
    "for line in infile:\n",
    "    line_as_list=line.strip().replace('\"',\"\").split(\",\")\n",
    "    taxdict[line_as_list[0]] = line_as_list[1:]\n",
    "\n",
    "infile.close()\n",
    "\n",
    "for t in taxids:\n",
    "    print t,\n",
    "#    print taxdict[t]\n",
    "    taxonomy=\"\"\n",
    "    for i in range(len(tax_levels)):\n",
    "        if taxdict[t][indices[i]-1] == \"\":\n",
    "            taxonomy+='unknown'+';'\n",
    "        else:\n",
    "            taxonomy+=taxdict[taxdict[t][indices[i]-1]][2]+\";\"\n",
    "    print taxonomy[:-1]\n",
    "    taxids_to_taxonomy[t] = taxonomy[:-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out *.tax file for SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "out=open(\"tax_for_SATIVA.tax\", 'w')\n",
    "\n",
    "for sp in taxon_to_recs:\n",
    "    for rec in taxon_to_recs[sp]:\n",
    "        out.write(\"%s\\t%s\\n\" %(rec,taxids_to_taxonomy[taxon_to_taxid[sp]]))\n",
    "        \n",
    "out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
